README.md
# pi (the number) on a raspberry pi (the computer) for pi-day (March 14, 2021)

Read file [HISTORY.md](../master/HISTORY.md) to see day by day notes.

Watch the [YouTube Video](https://youtu.be/Czo20LnoEDs)

Read the [/docs](https://cotarr.github.io/calc-pi-arm64-asm/)

My coding challenge is now complete.
You can see the result here: [pi-1000000-digit.txt](../master/output-example/pi-1000000-digit.txt)


The start date was Feb 14, 2021. Time needed is 1 month.
The coding challenge was to write a program in assembly language to calculate pi,
on a Raspberry Pi, by "Pi Day" March 14, 2021.
The personal goal was to learn a little about ARM64/arm-v8a assembly language.
The technical goal was to avoid use of any external math libraries.

### Models Used

- Raspberry Pi 4 Model B Rev 1.2
- Raspberry Pi 3 Model B Plus Rev 1.3
- Raspberry Pi 3 Model B Rev 1.2
- The 64 bit operating system does not load on Pi-2 and earlier.

### Operating System 64-Bit Raspberry Pi OS

Update: 2023-07-14

The 64 bit version of Raspberry Pi OS has been released as a stable release.
You can find it at https://www.raspberrypi.com/software/operating-systems/#raspberry-pi-os-64-bit
Please disregard any instructions to use the "beta" version.

### What am I using?

- Raspberry Pi 4 Model B Rev 1.2 / Raspberry Pi 3 Model B Plus Rev 1.3
- ARM64 (Beta) Debian GNU/Linux 10 (buster)
- GNU Assembler + GNU linker with arch=armv8-a and cpu=cortex-a72
- Editor: Atom 1.48 with package: Language-arm v2.1.1
- Editor settings: tab size 8 with hard tabs

### Check Model of Raspberry Pi (3B, 3B+, 4B)

```bash
~ $ cat /proc/cpuinfo | grep Model
Model		: Raspberry Pi 4 Model B Rev 1.2
~ $
```

### Check OS is "aarch64" with `lscpu`
```bash
~ $ lscpu

Architecture:        aarch64  <-------------- Required
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
Vendor ID:           ARM
Model:               3
Model name:          Cortex-A72
Stepping:            r0p3
CPU max MHz:         1500.0000
CPU min MHz:         600.0000
BogoMIPS:            108.00
Flags:               fp asimd evtstrm crc32 cpuid
~ $
```


### Installation on Raspberry Pi-3 or Pi-4

```bash
git clone https://github.com/cotarr/calc-pi-arm64-asm.git
cd calc-pi-arm64-asm
cd src

make
```


## Run from within the src folder

```bash
./calc-pi
```

- Expected command prompt is "Op Code:"
- Response (License an other intro not shown here)

```
I/O Initialized
Accuracy: 60 Digits (fraction part)
Variables initialized.
Op Code:
```

# walk through to calculate pi

This will use the following commands:

- `sigfigs 1000` will set accuracy to 1000 digits in base 10
- `c.pi` will calculate pi
- `print f` will print the number in page formated layout

This is step by step.

- Set accuracy to 1000 digits using `sigfigs` command

```
Op Code: sigfigs 1000
```
- Response

```
Accuracy: 1000 Digits (fraction part)
```

- Calculate pi with `c.pi` command

```
Op Code: c.pi
```
- Response

```
Calculating: Square Root 10005  (Elapsed: 0.008 Sec)
Calculating: Chudnovsky infinite series  (Elapsed: 0.018 Sec)

XREG   +3.14159265358979323846264338327950288419716939937510
YREG   +0.0
ZREG   +0.0
TREG   +0.0

  (Elapsed: 0.021 Sec)
```

- Print binary to decimal conversion in formatted text with `print f` command.

```
Op Code: print f
```
- Response

```
+3.
  1415926535 8979323846 2643383279 5028841971 6939937510 5820974944 5923078164 0628620899 8628034825 3421170679
  8214808651 3282306647 0938446095 5058223172 5359408128 4811174502 8410270193 8521105559 6446229489 5493038196
  4428810975 6659334461 2847564823 3786783165 2712019091 4564856692 3460348610 4543266482 1339360726 0249141273
  7245870066 0631558817 4881520920 9628292540 9171536436 7892590360 0113305305 4882046652 1384146951 9415116094
  3305727036 5759591953 0921861173 8193261179 3105118548 0744623799 6274956735 1885752724 8912279381 8301194912
  9833673362 4406566430 8602139494 6395224737 1907021798 6094370277 0539217176 2931767523 8467481846 7669405132
  0005681271 4526356082 7785771342 7577896091 7363717872 1468440901 2249534301 4654958537 1050792279 6892589235
  4201995611 2129021960 8640344181 5981362977 4771309960 5187072113 4999999837 2978049951 0597317328 1609631859
  5024459455 3469083026 4252230825 3344685035 2619311881 7101000313 7838752886 5875332083 8142061717 7669147303
  5982534904 2875546873 1159562863 8823537875 9375195778 1857780532 1712268066 1300192787 6611195909 2164201989
 (3809525720 )

  (Elapsed: 0.046 Sec)
```

## Commands available:

```
+ - * / . c.e c.pi c.pi.ch chs clrstk clrx
cmdlist D.vars D.fill D.ofst enter exit help
hex log logoff mmode prac print q recip
quit rcl rdown rup sf sigfigs sqrt sto
test version xy
```

## Chudnovsky Formula to compute pi

![CHudnovsky-Formula-Image](https://github.com/cotarr/calc-pi-arm64-asm/blob/main/images/Chudnovskyformula.jpg?raw=true)

## benchmark 2021-03-11 Calculation of pi

```
   Digits    Seconds
    10000      0.763
    20000      2.964
    50000     16.502
   100000     68.762
   200000    282.295
   500000   1694.834
  1000000   7153.571
```

![Calculation-Time-Image](https://github.com/cotarr/calc-pi-arm64-asm/blob/main/images/pi-calc-time.png?raw=true)


### Security Note

This application was intended for I/O limited to local keyboard and console output
within a Linux command line shell. This calculation includes a rather ubiquitous
use of memory pointers that have not been reviewed for safe pointer practices.
Therefore, modification of the program to service a direct internet connection
is not recommended.

System memory used for floating point number variables are defined in
math.s using statements to declare uninitialized blocks of memory
in the BSS section. These are statically allocated when the program is
started as part of the load image. No memory is dynamically allocated.

All input and output is performed using ARM64 `svc` system call statements.
All I/O functions are located in the "io_module.asm" file.
They are used to accept keyboard input, produce console output, capture
program text output to a disk file, and read the system clock.

There are no dependencies on third party libraries.


HISTORY.md
# Program history

### 2021-02-14 - Day 1 (28 days to go until pi day)

- Created empty git repository
- Installed 64 bit beta Raspberry OS from https://downloads.raspberrypi.org/raspios_lite_arm64/images/raspios_lite_arm64-2020-08-24/
- Put SD card into: Raspberry Pi 4 Model B Rev 1.2
- Updated packages to current version `apt-get update && apt-get upgrade`

- Check for 64 bit mode

```
$uname -m
aarch64
```
- Run lscpu to get cpu and arch

```
$lscpu
Architecture:        aarch64
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
Vendor ID:           ARM
Model:               3
Model name:          Cortex-A72
Stepping:            r0p3
CPU max MHz:         1500.0000
CPU min MHz:         600.0000
BogoMIPS:            108.00
Flags:               fp asimd evtstrm crc32 cpuid
```

- Created source file main.s iomodule.s, arch.inc, header.inc
- GNU assembler flags in arch.inc
```
  .arch armv8-a		// arm64 --> trying armv8-a
  .cpu  cortex-a72	// Raspberry Pi 4 from lscpu
```

- Wrote Hello World, print string with address in X0, first commit:

- main.s - assembler file with program entry point
- iomodule.s - Character print routines to stdout
- arch.s - Include file for 'cpu' and 'arch' directives
- header.s - Include file for configuration variables
- makefile - Build file to call assembler and linker

```
git checkout d5188777cda71522eb2428c5fdba4ab9a0a63314
```

### 2021-02-15 - Day-2 (27 days to go until pi day)

- Fixed several errors preserving registers in Hello World
- Added Keyboard input function KeyIn, reads line from console terminal with address in X0

```
git checkout be53bf079df40da133861fe7ebc44e38fe3a4d66
```

- util.s - New file with function to print byte and 64 bit word in hexadecimal format to stdout

```
git checkout 5ff998d0ff4f71e28e12d000c9aa12471d1471ab
```

- parser.s - New file to hold command parser code.

The command parser is a table of 8 byte null terminated strings and 8 byte jump addresses,
with 16 bytes per command. Upon a command match, the address of the handler is
put into the x30 register, and a `ret` command is executed as a "jump".
Initial commands: `cmdlist exit q quit test version`

Checks were added to make sure the stack pointer does not change and a check
that the command addresses are in byte alignment.

```
git checkout e54339d9f52c790d3e19fe42b6c0cc32b8e141d0
```

### 2021-02-16 - Day 3 (26 days to go until pi day)

Spent many hours reading asm64 docs. Following docs, cleaned up register use
following convention for register numbering. Cleaned up command parsing code in parser.s

- Added function ClearRegisters to zero all non-critical registers
- Added function PrintRegisters to show ARM registers x0 to x31
- Added function PrintFlags to show Z, C, N and V flags (may be inserted in loop)
- Added function to print 64 bit register as base-10 unsigned integer

```
git checkout 8b2540ab1f30d5fda32fc68e5f4d54e65f19bc59
```

### 2021-02-17 - Day 4 (25 days to go until pi day)

- practice.s - New file as sandbox to try learning ARM64 code.
- Added endian check for 32 bit w0 and 64 bit x0 memory load to register
- 64 bit string to integer converter

Use of command arguments: During user input an argument may be added
to a command followed by a space character to delimiter a string for parsing.
The address of the argument string is available in x0 in the command handler in parser.s.
Added a new function IntWordInput to convert integer string to binary unsigned 64 bit integer.
Error checking added for empty string and non-numeric characters.

```
git checkout fd759390c4cd42e42855eef002c2da5b597838f4
```

### 2021-02-18 - Day 5 (24 days to go until pi day)

- math.s - New file to hold variable declarations
- math-debug.s New file to hold debug tools used to write the program.
- arch-include.s - Renamed from arch.inc to get editor source highlighting.
- header-include.s - Renamed from header.inc to get editor source highlighting.

Typically, variables would be defined as binary floating point, with an
exponent, mantissa, and some extra words at the least significant end to absorb
errors which are called guard words.
```
   < binary exponent words>  <mantissa words> <guard words>
```

However, for the case of calculation of pi, scientific notation is not really
needed. To simplify the code, I am going to define the variables as fixed point.
In this case there must be a decimal point to separate integer part from fraction part.
```
   <binary integer part> /decimal point/ <binary fraction part>  <guard words>
```

In the case of this program, the variables are declared within the load image
of the application, rather than by dynamically allocated memory.

This raises the question of variable scope. Considering the names and references
to the variable area are complex, and size of the variables can be adjusted
by configuration variables, I intended to keep the scope of the name space
for variable definitions in the math.s module and then include arithmetic functions
as include files into the math.s file. This way I can have separate files
for math arithmetic functions, but all calculations will be within one object module.

- Added function FP_Initialize to setup variable space on program start

In order to view variables while writing the program, a tool is needed to display
variables in hexadecimal. This is useful when testing future code, such as integer addition.

- Add command with optional argument `hex [<reg-no>]` used to fully display one variable or abbreviate display all variables in hexadecimal

```
Op Code: hex
REG   Hand M.S. Word                      (no guard) L.S.W
ACC   ( 0) 0000000000000000 0000000000000000 0000000000000000 0000000000000000 .. 0000000000000000
OPR   ( 1) 0000000000000000 0000000000000000 0000000000000000 0000000000000000 .. 0000000000000000

src/help.s
/* -------------------------------------------------------------
	help.s

 This module contains user help

	Created:   2021-02-19
	Last Edit: 2021-03-08

----------------------------------------------------------------
MIT License

Copyright 2021 David Bolenbaugh

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
----------------------------------------------------------------
	Help
	Help_Welcome
------------------------------------------------------------- */

	.include "arch-include.s"	// .arch and .cpu directives
	.include "header-include.s"

	.global Help
	.global	Help_Welcome

// ------------------------------------------------------------
	.data
// ------------------------------------------------------------


/*---------------------------------------------------------------------

  Help Table

  8 character command for up to 7 character command zero terminated

  64 bit address pointer

--------------------------------------------------------------------*/

	.align 4
Help_Table:
	.ascii	"."
	.byte	0,0,0,0,0,0,0
	.quad	Help_print_dot

	.ascii	"+"
	.byte	0,0,0,0,0,0,0
	.quad	Help_plus_symbol

	.ascii	"-"
	.byte	0,0,0,0,0,0,0
	.quad	Help_minus_symbol

	.ascii	"/"
	.byte	0,0,0,0,0,0,0
	.quad	Help_slash_symbol

	.ascii	"*"
	.byte	0,0,0,0,0,0,0
	.quad	Help_star_symbol

	.ascii	"c.e"
	.byte	0,0,0,0,0
	.quad	Help_c_e

	.ascii	"c.pi"
	.byte	0,0,0,0
	.quad	Help_c_pi

	.ascii	"c.pi.ch"
	.byte	0
	.quad	Help_c_pi_ch

	.ascii	"clrstk"
	.byte	0,0
	.quad	Help_clrstk

	.ascii	"chs"
	.byte	0,0,0,0,0
	.quad	Help_chs

	.ascii	"clrx"
	.byte	0,0,0,0
	.quad	Help_clrx

	.ascii	"cmdlist"
	.byte	0
	.quad	Help_cmdlist

	.ascii	"D.fill"
	.byte	 0, 0
	.quad	Help_D_fill

	.ascii	"D.ofst"
	.byte	 0, 0
	.quad	Help_D_ofst

	.ascii	"D.vars"
	.byte	 0, 0
	.quad	Help_D_vars

	.ascii	"enter"
	.byte	0, 0, 0
	.quad	Help_enter

	.ascii	"exit"
	.byte	0, 0, 0, 0
	.quad	Help_exit

	.ascii	"hex"
	.byte	0, 0, 0, 0, 0
	.quad	Help_hex

	.ascii	"log"
	.byte	0, 0, 0, 0, 0
	.quad	Help_log

	.ascii	"logoff"
	.byte	0, 0
	.quad	Help_logoff


	.ascii	"mmode"
	.byte	0, 0, 0
	.quad	Help_mmode

	.ascii	"print"
	.byte	0,0,0
	.quad	Help_print_print

	.ascii	"q"
	.byte	0, 0, 0, 0, 0, 0, 0
	.quad	Help_q

	.ascii	"quit"
	.byte	0, 0, 0, 0
	.quad	Help_q

	.ascii	"rcl"
	.byte	0,0,0,0,0
	.quad	Help_rcl

	.ascii	"rdown"
	.byte	0,0,0
	.quad	Help_rdown

	.ascii	"recip"
	.byte	0,0,0
	.quad	Help_recip

	.ascii	"rup"
	.byte	0,0,0,0,0
	.quad	Help_rdown

	.ascii	"sf"
	.byte	0,0,0,0,0,0
	.quad	Help_sigfigs

	.ascii	"sigfigs"
	.byte	0
	.quad	Help_sigfigs

	.ascii	"sqrt"
	.byte	0,0,0,0
	.quad	Help_sqrt

	.ascii	"sto"
	.byte	0,0,0,0,0
	.quad	Help_sto

	.ascii	"xy"
	.byte	0,0,0,0,0,0
	.quad	Help_xy

Help_Table_End:
	.byte	0, 0, 0, 0, 0, 0, 0, 0
	.quad	0			// End of list


Help_plus_symbol:
	.ascii	"Usage: +\n\n"
	.ascii	"Description: Floating Point Addition\n"
	.ascii	"Parser:: XReg = YReg + XReg\n"
	.ascii	"Stack: YReg --> XReg, then roll stack down.\n"
	.byte	0

Help_minus_symbol:
	.ascii	"Usage: -\n\n"
	.ascii	"Description: Floating Point Subtraction\n"
	.ascii	"Parser: XReg = YReg - XReg\n"
	.ascii	"Stack: YReg --> XReg, then roll stack down.\n"
	.byte	0

Help_slash_symbol:
	.ascii	"Usage: /\n\n"
	.ascii	"Description: Floating Point Division\n"
	.ascii	"Parser: XReg = YReg / XReg\n"
	.ascii	"Stack: YReg --> XReg, then roll stack down.\n"
	.byte	0

Help_star_symbol:
	.ascii	"Usage: *\n\n"
	.ascii	"Description: Floating Point Multiplication\n"
	.ascii	"Parser: XReg = YReg * XReg\n"
	.ascii	"Stack: YReg --> XReg, then roll stack down.\n"
	.byte	0

Help_c_e:
	.ascii	"Usage: c.e\n\n"
	.ascii	"Description: Calculation of the constant e.\n"
	.ascii	"The calculation is performed by the summation of 1/n!.\n"
	.ascii	"Division loop calculates 1/n using ARM register 64/32 bit udiv/msub.\n"
	.byte	0

Help_c_pi:
Help_c_pi_ch:
	.ascii	"Usage: c.pi\n"
	.ascii	"Usage: c.pi.ch\n\n"
	.ascii	"Description: Calculation of the constant pi\n"
	.ascii	"The calculation is performed by the summation \n"
	.ascii	"involving the Chudnovsky Formula\n"
	.byte	0

Help_chs:
	.ascii	"Usage: chs\n\n"
	.ascii	"Description: Change Sign of XReg\n"
	.ascii	"Parser: xReg = Yreg / Xreg\n"
	.ascii	"Stack: YReg --> XReg, then roll stack down\n"
	.byte	0

Help_clrstk:
	.ascii	"Usage: clrstk\n\n"
	.ascii	"Description: Clear floating point stack to zero.\n"
	.ascii	"(XReg, YReg, Zreg, Treg)\n"
	.byte	0

Help_clrx:
	.ascii	"Usage: clrx\n\n"
	.ascii	"Description: Clear X, floating point XReg\n"
	.byte 	0

Help_cmdlist:
	.ascii	"Usage: cmdlist <optional first letter>\n\n"
	.ascii	"Description: Prints a list of availble commands. \n"
	.ascii	"To shorten the list provide the  first letter as a\n"
	.ascii	"command argument.\n"
	.byte	0


Help_D_fill:
	.ascii	"Usage: D.fill <handle>\n\n"
	.ascii	"Description: This is a debug command to fill a variable with\n"
	.ascii	"sequential byte value numbers. It starts with the integer part\n"
	.ascii	"0102030405060708 1011121314151617\n"
	.ascii	"next words in fraction part\n"
	.ascii	"2122232425262728 292A2B2C2D2E2F30\n"
	.ascii	"This is very useful to check low level functions, such as \n"
	.ascii	"shifting memory left or right 1 bit.\n"
	.byte	0

Help_D_ofst:
	.ascii	"Usage: D.ofst\n\n"
	.ascii	"Description: Show address offsets in hexadecimal\n"
	.byte	0

Help_D_vars:
	.ascii	"Usage: D.vars\n\n"
	.ascii	"Description: Show accuracy variables in hex and decimal\n"
	.byte	0

Help_enter:
	.ascii	"Usage: enter\n\n"
	.ascii	"Description: Duplicate X-Reg and roll X-->Y, Y-->Z, Z-->T (T is discarded)\n"
	.byte	0

Help_exit:
	.ascii	"Usage: exit\n\n"
	.ascii	"Description: Quit the program.\n"
	.byte	0

Help_hex:
	.ascii	"Usage: hex <optoinal variable handle number>\n\n"
	.ascii	"Description: Hex command is used to display variables in \n"
	.ascii	"binary (hexidecimal) format. If the hex command is called\n"
	.ascii	"without an argument, all the registers are printed showing\n"
	.ascii	"first the integer part followed by fraction part in\n"
	.ascii	"64-bit words in hexidecimal. If a variable hands is \n"
	.ascii	"is provided, the entire variable is printed.\n"
	.byte	0
Help_log:
	.ascii	"Usage: log\n\n"
	.ascii	"Description: The command 'log' starts a terminal log session.\n"
	.ascii	"The filenames are sequential as follows:\n"
	.ascii	"out/out001.txt, out/out002.txt, out/out003.txt ...\n"
	.ascii	"the folder 'out' is expected in the working directory.\n"
	.ascii	"Logging is stopped with the 'logoff' command\n"
	.byte	0

Help_logoff:
	.ascii	"Usage: logoff\n\n"
	.ascii	"Description: The command 'log' stops a terminal log session.\n"
	.byte	0

Help_mmode:
	.ascii	"Usage: mmode <optional integer bit pattern>\n\n"
	.ascii	"Descripton: Without argument, mmode displays MathMode variable.\n"
	.ascii	"   2   (0x04)  Force bitwise long division (shift and subtract)\n"
	.ascii	"   4   (0x04)  Disable: ARM 64 bit MUL/UMULH matrix multiplication\n"
	.ascii	"   8   (0x08)  Disable: ARM 32 bit UDIV/MSUB matrix division\n"
	.byte	0

Help_print_dot:
	.ascii	"Usage: .   (print X-Reg)\n\n"
	.ascii	"Usage: . f (formated - 10 and 1000 digit block output.\n"
//	.ascii	"Usage: . q (quiet - terminal output suppressed, 10 1000 format\n"
//	.ascii	"Usage: . u (unformatted - no line feed or page formatting\n"
	.ascii	"\nDescription: The . (period character) or 'print' will convert\n"
	.ascii	"the contents of X-reg from binary to decimail. The output\n"
	.ascii	"will be printed to stdout.\n"
	.ascii	"\nCommands '.' and 'print' are synonymous\n"
	.byte	0
Help_print_print:
	.ascii	"Usage: print   (print X-Reg)\n\n"
	.ascii	"Usage: print f (formated - 10 and 1000 digit block output.\n"
//	.ascii	"Usage: print q (quiet - terminal output suppressed, 10 1000 format\n"
//	.ascii	"Usage: print u (unformatted - no line feed or page formatting\n"
	.ascii	"\nDescription: The . (period character) or 'print' will convert\n"
	.ascii	"the contents of X-reg from binary to decimail. The output\n"
	.ascii	"will be printed to stdout.\n"
	.ascii	"\nCommands \".\" and \"print\" are synonymous\n"
	.byte	0

Help_q:
	.ascii	"Usage: quit\n\n"
	.ascii	"Description: Quit the program.\n"
	.byte	0

Help_rcl:
	.ascii	"Usage: rcl <reg-number>\n\n"
	.ascii	"Description: Recall storage register into XREG\n"
	.ascii	"Valid registers: 0 to "
	.byte	"TOP_REG_ASCII"
	.ascii	"\n"
	.ascii	"Stack Roll up: R-->X, X-->Y, Y-->Z, Z-->T\n"
	.byte	0

Help_rdown:
	.ascii	"Usage: rdown\n\n"
	.ascii	"Stack Roll DOWN: T-->Z, Z-->Y, Y-->X, X-->T\n"
	.byte	0

Help_recip:
	.ascii	"Usage: recip\n\n"
	.ascii	"Calculate reciprocal (1/x) of XReg\n"
	.ascii	"Stack not rotated after 1/x.\n"
	.byte	0

Help_rup:
	.ascii	"Usage: rup\n\n"
	.ascii	"Stack Roll up: X-->Y, Y-->Z, Z-->T, T-->X\n"
	.byte	0



Help_sqrt:
	.ascii	"Usage: sqrt\n\n"
	.ascii	"Calculate the square root of XReg\n"
	.ascii	"Stack not rotated after sqrt(XReg)\n"
	.byte	0

Help_sto:
	.ascii	"Usage: sto <reg-number>\n\n"
	.ascii	"Description: Store XREG into storage register\n"
	.ascii	"Valid registers: 0 to "
	.byte	"TOP_REG_ASCII"
	.ascii	"\n"
	.ascii	"Stack not rotated after sto()\n"
	.byte	0


Help_xy:
	.ascii	"Usage: xy\n\n"
	.ascii	"Description: Exchange XReg <> YReg.\n"
	.byte	0


Help_sigfigs:
	.ascii	"Usage: sf             (prints current accuracy)\n\n"
	.ascii	"Usage: sf   <integer> (set new accuracy digits base 10)\n"
//	.ascii	"Usage: sf w <integer> (set new accuracy 64 bit words)\n"
	.ascii	"Usage: sf e <integer> (set new extended digits, 0 for none)\n"
	.ascii	"Usage: sf v           (display accuracy verbose)\n"
//	.ascii	"Usage: sf K           (sets accuracy to 1K 1, 000 digits base 10)\n"
//	.ascii	"Usage: sf M           (sets accuracy to 1M 1, 000, 000 digits base 10)\n"
//	.ascii	"Usage: sf x           (sets accuracy to maximum)\n"
	.ascii	"\nThe sf (and sigfigs) commands aure used to set or display the current\n"
	.ascii	"precision level (significant digits) for floating point variables.\n"
	.ascii	"64 bit word size can be converted to base 10 number size by:\n"
	.ascii	"19.2659197224948 digit/QWord(64 bit) = log_base10(2^64)\n"
	.ascii	"Guard words provide additional precision to absorb rounding errors.\n"
	.ascii	"Set extended digits show result past specified accuracy.\n"
	.ascii	"Variable size INT_WSIZE, FCT_WSIZE and GUARDWORDS specified \n"
	.ascii	"in var_header.inc.\n"
	.byte	0

//=====================================
// Default help message
//=====================================
//
DefaultHelp:
	.ascii	"Usage: help <command name>\n\n"
	.ascii	"Description: help will provide description and\n"
	.ascii	"instructions for the use of a specific command.\n"
	.ascii	"To see a list of all commands, type 'cmdlist'. \n\n"
//	.ascii	"Help in html format is in the docs/ folder or on the web at \n"
//	.ascii      "https://cotarr.github.io/calc-pi-arm64-asm/\n\n"
	.ascii	"Help is available for the following commands:\n\n"
	.byte	0

HelpNotFound:
	.ascii	"No help was found for that command.\n\n"
	.ascii	"To see a list of all commands, type 'cmdlist'.\n"
	.ascii	"To see help for a specific command type: 'help <command>'.\n"
	.ascii	"Help is available for the following commands:\n\n"
	.byte	0

//
// Welcome message shown at program start
//
WelcomeMsg:
	.ascii	"SINGLE THREAD FLOATING POINT MULTI-PRECISION CALCULATOR\n\n"
	.ascii	"MIT License\n\n"
	.ascii	"Copyright 2021 David Bolenbaugh\n\n",
	.ascii	"Permission is hereby granted, free of charge, to any person obtaining a copy\n"
	.ascii	"of this software and associated documentation files (the \"Software\"), to deal\n"
	.ascii	"in the Software without restriction, including without limitation the rights\n"
	.ascii	"to use, copy, modify, merge, publish, distribute, sublicense, and/or sell\n"
	.ascii	"copies of the Software, and to permit persons to whom the Software is\n"
	.ascii	"furnished to do so, subject to the following conditions:\n\n"
	.ascii	"The above copyright notice and this permission notice shall be included in all\n"
	.ascii	"copies or substantial portions of the Software.\n\n"
	.ascii	"THE SOFTWARE IS PROVIDED \"AS IS\" WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\n"
	.ascii	"IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, \n"
	.ascii	"FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\n"
	.ascii	"AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\n"
	.ascii	"LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, \n"
	.ascii	"OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE\n"
	.ascii	"SOFTWARE.\n\n"
	.ascii	"Source code: https://github.com/cotarr/calc-pi-arm64-asm\n\n"

	.ascii	"Calculation of Pi on Raspberry Pi\n"
	.ascii	"Written in GNU Assembler (as)\n"
.ifdef CORTEXA72
	.ascii	"Assembled arch=armv8-a cpu=cortex-a72\n"
.endif
.ifdef	CORTEXA53
	.ascii	"Assembled arch=armv8-a cpu=cortex-a53\n"
.endif
	.ascii	"\n"
	.byte	0
	.align 4

// ------------------------------------------------------------
	.text
	.align 4
// ------------------------------------------------------------


/* -----------------------------------------------------------------------------

    Help Interface

    Input: x0 address to input command buffer

     Output: none

----------------------------------------------------------------------------- */
Help:
	sub	sp, sp, #64		// Reserve 8 words
	str	x30, [sp, #0]		// Preserve these registers
	str	x29, [sp, #8]
	str	x0,  [sp, #16]		// Command line argument pointer
	str	x1,  [sp, #24]
	str	x9,  [sp, #32]		// Scratch
	str	x10, [sp, #40]		// Index
	str	x11, [sp, #48]		// Argument address pointer
	str	x12, [sp, #56]		// Table address pointer

	bl	CROut
	mov	x11, x0			// Save pointer to command line argument
//
// Check alignment of command table (could be entry/code error)
//
	ldr	x0, =Help_Table_End	// Check for byte alignment
	ands	x0, x0, #0x0f		// Should be zero
	b.eq	10f
	ldr	x0, =Command_Error2	// Error \message pointer
	mov	x1, #2732		// 12 bit error code
	b	FatalError
10:
//
// Check for help argument, if no argument then show default help
	cmp	w11, #0			// Is argument pointer = 0?
	b.ne	20f			// No, parse argument
	ldr	x0, =DefaultHelp	// Else, yes zero, no argument, default help
	bl	StrOut
	bl	PrintHelpList
	b.al	help_exit
20:
	ldr	x12, =Help_Table	// Initialize address pointer to Table Address
//
// Loop here for next help command check
//
Help_next_cmd:
	ldr	x0, [x12]		// Check for past last command in table
	cmp	x0, #0
	b.ne	30f			// Zero marker found, end of command table
	ldr	x0, =HelpNotFound
	bl	StrOut
	bl	PrintHelpList
	b.al	help_exit

30:
	mov	x10, #0			// Reset pointer to start of record
//
// Loop here for next character check
//
Help_next_char:
	ldrb	w0, [x11, x10]
	ldrb	w9, [x12, x10]
	cmp	w0, w9
	b.ne	Help_Tab_Next
	cmp	x10, #8			// Only 7 char + zero allowed
	b.ne	40f			// 8 Found is fatal error in table
	ldr	x0, =Command_Error1	// Error message pointer
	mov	x1, #2455		// 12 bit error code
	b	FatalError
40:
	add	X10, X10, #1
	ldrb	w9, [x12, x10]
	cmp	w9, #0			// No more characters?
	b.ne	Help_next_char		// Not zero, more to check
	ldrb	w0, [x11, x10]
	cmp	w0, #0x20		// ascii space
	b.eq	Help_Tab_Match
	cmp	w0, #0x00		// ascii space
	b.eq	Help_Tab_Match
	b.al	Help_Tab_Next			// Not match zero or space on next char
Help_Tab_Match:
	mov	x10, #8
	ldr	x0, [x12, x10]
	bl	StrOut
	b.al	help_exit
Help_Tab_Next:
	add	x12, x12, #16		// Point to next command
	b.al	Help_next_cmd
help_exit:
	bl	CROut

	ldr	x30, [sp, #0]		// restore registers
	ldr	x29, [sp, #8]
	ldr	x0,  [sp, #16]
	ldr	x1,  [sp, #24]
	ldr	x9,  [sp, #32]
	ldr	x10, [sp, #40]
	ldr	x11, [sp, #48]
	ldr	x12, [sp, #56]
	add	sp, sp, #64
	ret

Command_Error1:
	.ascii	"HelpCmd: zero end marker not found in text table"
	.byte	0
Command_Error2:
	.ascii	"HelpCmd: End of help table not QWord aligned, probably table error"
	.byte 	0
	.align 4
/* ----------------------------------------------

 Print Command List

 Input:

--------------------------------------------- */

PrintHelpList:
	sub	sp, sp, #80		// Reserve 10 words
	str	x30, [sp, #0]		// Preserve these registers
	str	x29, [sp, #8]
	str	x0,  [sp, #16]
	str	x9,  [sp, #24]
	str	x10, [sp, #32]
	str	x11, [sp, #40]

	mov	x9, #0			// Index to character
	mov	x11, #0			// Line feed counter
	ldr	x10, =Help_Table	// Address of command table
loop1:
	ldrb	w0, [x10, x9]		// Get character to see if done
	cmp	w0, #0			// Is it zero, then done
	b.eq	done
	mov	x0, X10			// address of command string in table
	bl	StrOut			// Print command
	mov	x0, #0x020		// ascii space
	bl	CharOut
//	add	x11, x11, #1			// Increment line feed counter
//	cmp	x11, #3			// Check limit
//	bl	PrintFlags
//	b.	.skip2
//	mov	r15, 0			// Reset counter
//	call	CROut			// Return + line feed
skip2:
	add	x10, x10, #16		// increment to next work
	mov	x9, #0
	b.al	loop1
done:
	bl	CROut

	ldr	x30, [sp, #0]		// restore registers
	ldr	x29, [sp, #8]
	ldr	x0,  [sp, #16]
	ldr	x9,  [sp, #24]
	ldr	x10, [sp, #32]
	ldr	x11, [sp, #40]
	add	sp, sp, #80
	ret

// TODO print all help

Help_Welcome:
	sub	sp, sp, #32		// Reserve 4 words
	str	x30, [sp, #0]
	str	x29, [sp, #8]
	str	x0,  [sp, #16]

	ldr	x0, =WelcomeMsg
	bl	StrOut

	ldr	x30, [sp, #0]		// Restore registers
	ldr	x29, [sp, #8]
	ldr	x0,  [sp, #16]
	add	sp, sp, #32
	ret

	.end
